flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -646,14 +658,52 @@ private TaskManagerSlot createAndRegisterTaskManagerSlot(SlotID slotId, Resource
 	@Nullable
 	private PendingTaskManagerSlot findExactlyMatchingPendingTaskManagerSlot(ResourceProfile resourceProfile) {
 		for (PendingTaskManagerSlot pendingTaskManagerSlot : pendingSlots.values()) {
-			if (pendingTaskManagerSlot.getResourceProfile().equals(resourceProfile)) {
+			if (pendingSlotExactlyMatchingResourceProfile(pendingTaskManagerSlot, resourceProfile)) {
 				return pendingTaskManagerSlot;
 			}
 		}
 
 		return null;
 	}
 
+	private boolean pendingSlotExactlyMatchingResourceProfile(PendingTaskManagerSlot pendingTaskManagerSlot, ResourceProfile resourceProfile) {
+		return pendingTaskManagerSlot.getResourceProfile().equals(resourceProfile);
+	}
+
+	private boolean totalSlotNumExceedMaxLimitAfterRegister(SlotReport slotReport) {
+		final int numReportedNewSlots = slotReport.getNumSlotStatus();
+		final int numRegisteredSlots =  getNumberRegisteredSlots();
+		final int numPendingSlots = getNumberPendingTaskManagerSlots();
+
+		// check if the total number exceed before matching pending slot.
+		if (numRegisteredSlots + numPendingSlots + numReportedNewSlots <= maxSlotNum) {
+			return false;
+		}
+
+		// check how many exceed slots could be consumed by pending slot.
+		final int totalSlotNum = numRegisteredSlots + numPendingSlots + getNumNonPendingReportedNewSlots(slotReport, numReportedNewSlots);
+		if (totalSlotNum > maxSlotNum) {
+			return true;
+		}
+
+		return false;
+	}
+
+	private int getNumNonPendingReportedNewSlots(SlotReport slotReport, int numReportedNewSlots) {
```

Comment:
Is `numReportedNewSlots` always the same as `slotReport.getNumSlotStatus()`?

Reply:
Yes, I pass this param to avoid calling `slotReport.getNumSlotStatus()` multiple times.

Reply:
> Yes, I pass this param to avoid calling `slotReport.getNumSlotStatus()` multiple times.

Why? Are there performance concerns?

Reply:
I think we could gain performance benefits from it. But true, the benefit is trivial. I would remove this param.